### ISSUE TITLE
Raspberry Pi 4: Can't mount SD card during boot

### ISSUE THREAD

**Issue by cwyc** [NixOS/nixpkgs#135828]
### Describe the bug
The generic aarch64 SD image (https://hydra.nixos.org/build/150582927) won't boot on my Raspberry Pi 4B 8GB. 

When powered on, it progresses past U-Boot into NixOS Stage 1. 
There, it prints the following error message about every second:
```
mmc1: ADMA error: 0x02000000
mmc1: sdhci: =========== SDHCI REGISTER DUMP ============
mmc1: sdhci: Blk size:  0x00008008 | Blk cnt:  0x00000001
mmc1: sdhci: Sys addr:  0x00000000 | Version:  0x00001002
mmc1: sdhci: Argument:  0x00000000 | Trn mode: 0x00000013
mmc1: sdhci: Present:   0x1fff0206 | Host ctl: 0x00000011
mmc1: sdhci: Power:     0x0000000f | Blk gap:  0x00000080
mmc1: sdhci: Wake-up:   0x00000000 | Clock:    0x0000fa07
mmc1: sdhci: Timeout:   0x00000000 | Int stat: 0000000000
mmc1: sdhci: Int enab:  0x03ff100b | Sig enab: 0x03ff100b
mmc1: sdhci: ACmd stat: 0x00000000 | Slot int: 0x00000000
mmc1: sdhci: Caps:      0x45ee6432 | Caps_1:   0x0000a525
mmc1: sdhci: Cmd:       0x0000333a | Max curr: 0x00080008
mmc1: sdhci: Resp[0]:   0x00000920 | Resp[1]:  0x00e6cf7f
mmc1: sdhci: Resp[2]:   0x325b5900 | Resp[3]:  0x00400e00
mmc1: sdhci: Host ctl2: 0x00000008
mmc1: sdhci: ADMA Err:  0x00000001 | ADMA Ptr: 0xc3000200
mmc1: sdhci: ============================================
mmc1: sdhci: c3000200: DMA 0xf3403000, LEN 0x0008, Attr=0x21
mmc1: sdhci: c3000208: DMA 0x00000000, LEN 0x0000, Attr=0x03
mmc1: error -5 whilst initialising SD card
```
And it finally fails mounting the root partition:
```
Timed out waiting for device /dev/disk/by-label/NIXOS_SD, trying to mount anyway.
mounting /dev/disk/by-label/NIXOS_SD on /...
mount: mounting /dev/disk/by-label/NIXOS_SD on /mnt-root/ failed: No such file or directory

An error occured in stage 1 of the boot process, which must mount the
root filesystem on '/mnt-root' then start stage 2. Press one
of the following keys:

  r) to reboot immediately
  *) to ignore the error and contine
```

This happens with `sd_image` and `sd_image_new_kernel` images, and I've also tried building it with `boot.kernelPackages = pkgs.linuxPackages_rpi4;`.
However, Raspberry Pi OS boots successfully when flashed onto the same SD card. Using that, I've run `rpi-update`, which has not fixed the problem.

### Steps To Reproduce
1. Download and decompress https://hydra.nixos.org/build/150582927
2. Flash it onto an SD card
3. Insert it into a Raspberry Pi 4B, and power it on

### Expected behavior
It should boot to the login screen.

### Additional context
These links might be relevant:
https://www.raspberrypi.org/forums/viewtopic.php?t=314845

> I think the problem is that the CM4, like newer Pi 4s, has a BCM2711C0. They differ in that the EMMC2 interface is on the ARM (not VPU) memory bus, so the (dma-)ranges property is different.

https://archlinuxarm.org/forum/viewtopic.php?t=15422&p=67103

### Notify maintainers

@samueldr 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

(n/a, can't boot)

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module: 
  - nixos/modules/installer/sd-card/sd-image-aarch64.nix
```

**Comment by cwyc** [comment]
Unfortunately, my Raspberry Pi just died (the hw is terribly unreliable!), so I won't be able to test any fixes.

**Comment by alyssais** [comment]
I have the same problem.

**Comment by superherointj** [comment]
@alyssais Can you include more information on your scenario? RPi4 Chipset? Is it BCM2711C0? SDCard used? Steps done. Etc.

**Comment by superherointj** [comment]
`newer Pi 4s, has a BCM2711C0. They differ in that the EMMC2 interface is on the ARM (not VPU) memory bus, so the (dma-)ranges property is different. The firmware takes care of the changes, but by bypassing it you miss out.`
```
-- PhilE - Raspberry Pi Engineer

```
you are right. Able to boot after changing the dma-range for emmc2bus node.
Modified the dma-range as below:
dma-ranges = <0x00000000 0x00000000 0x00000000 0x00000000 0xfc000000>;
```
-- parthitce

**Comment by alyssais** [comment]
I've tried multiple different SD cards from different manufacturers.  I've been building images with <nixpkgs/nixos/modules/installer/sd-card-aarch64.nix> (and the newkernel variants) from various different versions of Nixpkgs, all of which have produced the same error.  I've also tried with the Raspberry Pi 4 kernel in Nixpkgs.

<details>
<summary>/proc/cpuinfo</summary>
processor	: 0
BogoMIPS	: 108.00
Features	: fp asimd evtstrm crc32 cpuid
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0xd08
CPU revision	: 3

processor	: 1
BogoMIPS	: 108.00
Features	: fp asimd evtstrm crc32 cpuid
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0xd08
CPU revision	: 3

processor	: 2
BogoMIPS	: 108.00
Features	: fp asimd evtstrm crc32 cpuid
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0xd08
CPU revision	: 3

processor	: 3
BogoMIPS	: 108.00
Features	: fp asimd evtstrm crc32 cpuid
CPU implementer	: 0x41
CPU architecture: 8
CPU variant	: 0x0
CPU part	: 0xd08
CPU revision	: 3

Hardware	: BCM2835
Revision	: d03114
Serial		: 10000000fd4d9bb3
Model		: Raspberry Pi 4 Model B Rev 1.4
</summary>

**Comment by K900** [comment]
It seems removing `FDTDIR` from the bootloader config is enough to get something to boot, which to me implies that the DTB passed on by the bootloader works fine, but the DTBs NixOS generates are wrong.

**Comment by K900** [comment]
Applying the `emmc2bus` change from the running DT makes things work fine. Here it is as an overlay:

```dts
// SPDX-License-Identifier: GPL-2.0
/dts-v1/;
/plugin/;
/ {
    compatible = "brcm,bcm2711";
    fragment@1 {
        target = <&emmc2bus>;
        __overlay__ {
            dma-ranges = <0x00 0x00 0x00 0x00 0xfc000000>;
        };
    };
};
```

There's also an interesting note in the device tree: https://github.com/raspberrypi/linux/blob/rpi-5.10.y/arch/arm/boot/dts/bcm2711.dtsi#L423

**Comment by K900** [comment]
So I guess the correct solution here would be to just build the NixOS overlays into a separate file, then tell u-boot to load that on top of the device tree passed in by the original bootloader?

**Comment by superherointj** [comment]
> So I guess the correct solution here would be to just build the NixOS overlays into a separate file, then tell u-boot to load that on top of the device tree passed in by the original bootloader?

@samueldr ^

**Comment by samueldr** [comment]
(Answering for the DT issues)

@K900 your link isn't where your overlay values are being set:

 - https://github.com/raspberrypi/linux/blob/4a929be607a35b33f4d56a77e4ac70613b674e2c/arch/arm/boot/dts/bcm2711.dtsi#L435
 
```
		dma-ranges = <0x0 0xc0000000  0x0 0x00000000  0x40000000>;
```

Your overlay value sets it at:

```
            dma-ranges = <0x00 0x00 0x00 0x00 0xfc000000>;
```

How did you get the values you're using?

I see [the pre-built dtb](https://github.com/raspberrypi/firmware/blob/e3234b68ec227d0fa8e3f31c56f1370b1c2af880/boot/bcm2711-rpi-4-b.dtb) uses the same as the vendor kernel and mainline does:

```
dma-ranges = <0x00 0xc0000000 0x00 0x00 0x40000000>;
```

(verified throuch `dtc --sort ...`.)

So, *something* knows to affect the `dma-ranges` property of the `emmc2bus`, and we'd need to figure our what is, and why it is, so we can then ask around at mainline (where?) about a proper solution I guess.

In *theory*, if your values comes from the running U-Boot, U-Boot should have loaded the `.dtb` found on the SD card image, which in turn should be coming from the `raspberrypi/firmware` repo. Though maybe it's not loading the `.dtb` file as I expected?

Now, if it's coming from a non-U-Boot boot (or if U-Boot doesn't load the .dtb file as I assumed) then it means the *vendor firmware* itself is setting it to that value. Which might mean something is wrong in those `dts` files. But it would surprise me that the foundation's device tree is wrong. Especially since @alyssais tested against `linux_rpi4` AFAIUI.

**Comment by samueldr** [comment]
*Now that I've read that forum post*...

> I think the problem is that the CM4, like newer Pi 4s, **has a BCM2711C0.**

Sounds like a hardware revision that isn't being accounted for in mainline, and maybe not even in the vendor kernel, and instead being handled entirely on their firmware? **UGH!**

**Comment by K900** [comment]
The comment in my link says that the firmware is supposed to patch the DT at boot time (why??!), and in my testing it does exactly that. The values in the overlay are what I got from the firmware patched DT as loaded by Raspbian.

**Comment by samueldr** [comment]
Yes, sorry, as you might now see I updated the post since I saw the source. And I hate every ounce of it.

**Comment by K900** [comment]
Definitely +1 on the hate. Also, from my limited digging, this is handled in the proprietary Broadcom code, so we're basically stuck with it forever...

**Comment by K900** [comment]
Oh, and I've tested and this affects both the mainline kernel and the Foundation kernel.

**Comment by samueldr** [comment]
Yes, without testing (I can't not having a newer pi4), I can say I see it's likely 100% that.

What will need to happen is that *somehow* U-Boot will need to be aware of properties to "forward" from the "firmware-native" DT.

**Comment by K900** [comment]
It might be more viable to forward the entire DT, and then apply overlays as real overlays at boot time... Recent uboot versions should support that, we'll just need to do some changes to bootloader config generation...

**Comment by K900** [comment]
Relevant uboot change, for reference: https://github.com/u-boot/u-boot/commit/69076dff2284ed099cc0583e5e64bd8012d1ab5c#diff-af77874a271c64e907178ed13c27a3de60e4cfb5a9da7d092796c6329ed62d04

**Comment by samueldr** [comment]
(That overlay change wouldn't work really, as it would break older BCM2711B0 boards AFAIUI.)

**Comment by samueldr** [comment]
> It might be more viable to forward the entire DT, and then apply overlays as real overlays at boot time

*If only it was this simple*. We need to support *one boot flow*, or get much much more resources if we need to start supporting more boot flows.

The boot flow that was decided upon by predecessor decision makers, and is the preferred boot flow by mainline U-Boot is the [Generic Distro Configuration Concept](https://source.denx.de/u-boot/u-boot/-/blob/v2021.07/doc/README.distro). This is what boots *all mainline-supported boards* through our universal images. Whether it is the SD card, or transitively the UEFI iso.

Now, this was actually tangential to the problem, but it sets the scene for what comes here: we need to load the *kernel provided* device tree binaries. While mainline states that it's not necessary, the reality is that it is necessary to load the device tree binaries from the same revision the kernel was built from. Mainline fixes *too many things* in their opinion about the device trees.

**Comment by K900** [comment]
The overlay change I posted will definitely not work upstream - a proper solution has to involve the firmware provided DT somehow. It does, however, work as a temporary override for people like myself or Alyssa who have a known C0 stepping board.

**Comment by K900** [comment]
> > It might be more viable to forward the entire DT, and then apply overlays as real overlays at boot time
> 
> _If only it was this simple_. We need to support _one boot flow_, or get much much more resources if we need to start supporting more boot flows.
> 
> The boot flow that was decided upon by predecessor decision makers, and is the preferred boot flow by mainline U-Boot is the [Generic Distro Configuration Concept](https://source.denx.de/u-boot/u-boot/-/blob/v2021.07/doc/README.distro). This is what boots _all mainline-supported boards_ through our universal images. Whether it is the SD card, or transitively the UEFI iso.
> 
> Now, this was actually tangential to the problem, but it sets the scene for what comes here: we need to load the _kernel provided_ device tree binaries. While mainline states that it's not necessary, the reality is that it is necessary to load the device tree binaries from the same revision the kernel was built from. Mainline fixes _too many things_ in their opinion about the device trees.

Yeah, I can see how this can be a problem. The other option is an uboot level hack that only applies to Pi4 models, I guess... And even then the hack will have to stay in sync with mainline if the aliases ever change... Jesus fuck, Broadcom.

**Comment by samueldr** [comment]
> Jesus fuck, Broadcom.

For once the problem actually comes from the tradition of assuming the kernel knows best about the hardware's own description. The firmware is doing the right thing: describing its own hardware.

The kernel's view about device trees should, in my opinion, always have been built as discrete overlays on top of the firmware-innate FDT, rather than wholesale overriding it.

It would also have made loading overlays *by the firmware* automagically work. E.g. those raspberry pi overlay bords with an on-board DTB overlay that the firmware loads.

**Comment by K900** [comment]
Also, here's another thing I haven't actually looked into: it's entirely possible the firmware patches more than just this one node, and it's technically entirely possible for them to start patching even more nodes in a firmware update, as Raspbian (and seemingly most others) just trust the firmware provided DT.

**Comment by samueldr** [comment]
Indeed, it already does in somesituations!

**Comment by K900** [comment]
Does anyone have a B0 stepping board on hand they can dump the DT from? I can upload my own dump from a C0 tomorrow morning, and then diff them.

**Comment by VergeDX** [comment]
Same issues here, I have a rpi 4b with the latest firmware.
I'm willing to help : )

**Comment by K900** [comment]
Does adding the overlay in https://github.com/NixOS/nixpkgs/issues/135828#issuecomment-918359063 to `hardware.deviceTree.overlays` help?

**Comment by VergeDX** [comment]
@K900 I'm trying to. I'm using this file and the latest nixos-21.05 channel.
https://github.com/VergeDX/config-nixpkgs/blob/d8e8abdcd856a8f5b89d970bbd4d9a7944d81b5f/rpi/sd-image.nix

**Comment by VergeDX** [comment]
@K900 Perfect! That's fixes this issues.

**Comment by anirrudh** [comment]
@K900 YOU ARE THE BEST! I can confirm this works!

**Comment by K900** [comment]
So uh. I finally got some time to investigate this. I got a bunch of dumps from other people's boards running different distros/firmwares/dtc versions, and my god the DTs are all over the place. It seems like a better approach would be to compare two boards running the exact same software. Unfortunately, I can't get my hands on a B0 stepping board :( Is there anyone with an older board here who would be willing to flash, say, the latest Raspbian image from the website, dump the DT and compare notes?

**Comment by samueldr** [comment]
@K900 I really don't want to imply you're doing something you don't need to. But I'm curious as to why you're comparing the DTs that way.

Given your workaround works, I'm 99.9% sure you found the actual problem. I guess the only other relevant thing here is that there could be other FDT nodes that the firmware modifies when handling the dtb file?

Anything else I'm missing?

I want to stress out that **I want to hear your thoughts about that**.

As for myself, I think the solution will be to make U-Boot aware that on the bcm2711 Raspberry Pi family of boards it needs to transmit the innate `dma-range` property for the `&emmc2bus` node to any new FDT. And sadly **any softwarey/firmwarey bit that loads a new FDT will need to somehow know about that vendor-specific quirk**.

Alternatively, the kernels could be made aware of that deficiency with a kind of quirk, but that won't help with non-Linux OS.

The *actual* solution would be to never load a full replacement for the innate FDT, but that's not how mainline Linux is designed.

**Comment by K900** [comment]
> I guess the only other relevant thing here is that there could be other FDT nodes that the firmware modifies when handling the dtb file?

Yep, exactly this. There are a few references to the firmware modifying "a lot" or "a bunch" of things in the DT before boot by the Pi Foundation engineers, so I think we should at least be aware of all of them. 

> As for myself, I think the solution will be to make U-Boot aware that on the bcm2711 Raspberry Pi family of boards it needs to transmit the innate dma-range property for the &emmc2bus node to any new FDT. And sadly any softwarey/firmwarey bit that loads a new FDT will need to somehow know about that vendor-specific quirk.

That is the approach I'm hoping to take as well, but before we do that, I really want to make sure nothing else in the DT is affected in a meaningful way. 

> The actual solution would be to never load a full replacement for the innate FDT, but that's not how mainline Linux is designed.

Notably the FDT isn't actually _innate_ here - it's loaded from the firmware partition, patched by the bootloader, then passed back into the kernel. We could potentially abuse this in horrible ways - e.g. have a custom uboot that writes the right DT to the right place and then reboots the board to let the firmware patch it?

**Comment by samueldr** [comment]
> Notably the FDT isn't actually *innate* here [...]

Yeah, I know there's a dtb file loaded from the firmware partition, but for all intents and purposes for further stages, I think *it has to be assumed it is the innate FDT*. The fact the vendor firmware uses dtb files to populate the innate FDT is an implementation detail.

> have a custom uboot that writes the right DT to the right place and then reboots the board to let the firmware patch it?

*/me shudders*

I *think* U-Boot needs the vendor-flavoured device tree to be loaded at the FDT for the Raspberry Pi 4 to work. So this wouldn't really work for mainline-flavoured device trees; U-Boot wouldn't load after setting it so that the "innate" FDT is mainline-flavoured.

Somewhat relatedly, I wonder if, and how, the Tianocore builds are affected. I don't remember if they rely strictly on the "innate" FDT. I guess further stages (e.g. GRUB) would break the FDT if they load a dtb file, mirroring concerns about all softwarey/firmwarey bits needing to be aware.

* * *

As for comparing DTs, two tips:

 - Prepare a "good" disk image with raspbian, one which is already prepared with dtc installed, so you can get the same exact system booted on all user's hardware.
 - Use `dtc --sort /proc/device-tree > dumped.dts`; though I assume this is what you were doing already, hopefully.

If you provide a "starter" .img file I'll try on my 2GB and 4GB early revisions.

**Comment by K900** [comment]
TianoCore was actually a good pointer: https://github.com/tianocore/edk2-platforms/commit/bad0b9134b0e4a4e60d083c3d697d5ad99c2a7e9#diff-a4c80030ebbe12ded482507369d24076dbca1e8c3a42d0422f63958f2c93f86f

So at least they weren't checking the SoC revision before this change, and it looks like there are no more conditional changes after that. 

Assuming the Pi Foundation officially supports this, it looks like a uboot level hack should be enough here?

**Comment by K900** [comment]
Too bad we can't just use TianoCore and punt off responsibility... (not without breaking people's setups anyway)

**Comment by samueldr** [comment]
> Too bad we can't just use TianoCore and punt off responsibility...

We ca...

> (not without breaking people's setups anyway)

oh... right...

Though this does kind of show I should continue pushing forward in [making UEFI the default happy path on ARM](https://nixos.wiki/wiki/NixOS_on_ARM/UEFI).

* * *

> So at least they weren't checking the SoC revision before this change

The change is for the ACPI tables; they're doing the equivalent change that a DT-based solution would.

> Assuming the Pi Foundation officially supports this

They don't AFAIK

> it looks like a uboot level hack should be enough here?

Yes, it shows that U-Boot can either

 - Forward the node value from the "innate" DT
 - Hardcode the possible node values and always apply them

Though, now that I think about it, I don't know if the FDT loading has any way for platforms quirks to hook themselves. That will be a "fun" behaviour to implement.

**Comment by K900** [comment]
> Though this does kind of show I should continue pushing 
forward in making UEFI the default happy path on ARM.

Please do, and best of luck. (absolutely zero sarcasm intended, just in case)

> The change is for the ACPI tables; they're doing the equivalent change that a DT-based solution would.

Yeah, I know. My point isn't that they do this, my point is that *they don't do anything else*, which would indicate this change is enough.

**Comment by K900** [comment]
Also, looking at https://github.com/pftf, it seems like Broadcom, ARM and Pi Foundation folks are somehow involved with this, so it should be decently safe to assume some level of insider knowledge?

**Comment by K900** [comment]
Also looks like uboot actually already does some DT patching on the Pi? I'm not sure when exactly that code runs though... 
https://github.com/u-boot/u-boot/blob/master/board/raspberrypi/rpi/rpi.c#L498

**Comment by samueldr** [comment]
:eyes: https://source.denx.de/u-boot/u-boot/-/blob/v2021.07/include/fdt_support.h#L176-181

**Comment by K900** [comment]
I guess the next step would be to actually hack this into uboot... 

The tricky part is probably going to be reading the SoC revision, but I could try just hardcoding it tomorrow for the sake of experimentation .

**Comment by samueldr** [comment]
Hints:

 - https://www.raspberrypi.org/documentation/computers/raspberry-pi.html#new-style-revision-codes
 - https://www.raspberrypi.org/documentation/computers/raspberry-pi.html#new-style-revision-codes
 - https://source.denx.de/u-boot/u-boot/-/blob/v2021.07/board/raspberrypi/rpi/rpi.c#L451-456

Comparing some Raspberry Pi 4B identifiers found:

```
  "NOQuuuWuF MMM CCCC PPPP TTTTTTTT RRRR"
0xa03111: "1 010 0000 0011 00010001 0001" 
0xc03111: "1 100 0000 0011 00010001 0001"
0xc03112: "1 100 0000 0011 00010001 0010"
0xd03114: "1 101 0000 0011 00010001 0100"
```

I think here we have that info in the Revision fields. The manufacturers shouldn't matter (they won't), and everything else is the same.

In that U-Boot link they're reading the Type fields.

A new data point would need to be saved for the revision. This data point can then be used in `ft_board_setup`.

Another required patch will be handling of `ft_board_setup` in the generic distro boot, separate but intrinsically linked for our needs.

**Comment by K900** [comment]
The revision might not be enough, at least I can't find any definite information saying 1.4 boards only use C0 SoCs... 

We should be able to just read the same register TianoCore does though. 

I'm still trying to figure out if ft_board_setup is called in the generic boot process, but honestly it might be easier to just look at the log once I get the board hooked up to a display again. 

Either way, I should probably go sleep now and figure this out tomorrow.

**Comment by samueldr** [comment]
> I'm still trying to figure out if ft_board_setup is called in the generic boot process, but honestly it might be easier to just look at the log once I get the board hooked up to a display again.

Oh, looked again at the code and it *might* be called; `image_setup_libfdt` uses it, which is used in `bootm` in `boot_setup_fdt`. Looks likely it's used in the end.

**Comment by samueldr** [comment]
Looks like U-Boot may know already about the DMA range changes... (duh... U-Boot works...)

https://lore.kernel.org/u-boot/20210112125531.26547-1-nsaenzjulienne@suse.de/

It seems you might be able to "just" always set the dma range properties for `&emmc2` in `ft_board_setup`?

**Comment by K900** [comment]
For reference, here's the entire series in a slightly more readable UI: https://patchwork.ozlabs.org/project/uboot/list/?series=223890&state=*

Unfortunately, it looks like it's not reading the data we need...

**Comment by K900** [comment]
OK, it is actually reading the data we need, we just can't get to it in a reasonable way - there's no good place to get a reference to the MMC device. It might actually be easier to just copy the value from the innate DT... anyway, that's definitely going to have to wait until tomorrow.

**Comment by K900** [comment]
So the U-Boot hack worked. I've pushed it here: https://github.com/K900/u-boot/commit/8fac20e36ba09e31e2a2988342a5fa7897f6c2b8 and I'm not sure how to get it upstreamed now (or if we even should).

I've also uploaded a patched U-Boot binary, built from latest nixos-unstable with this patch applied, to https://static.0upti.me/pi4/u-boot-rpi4.bin (or of course, you're welcome to build your own). You should be able to simply drop it into your firmware partition and boot without any sort of device tree hacks on the OS side.

**Comment by K900** [comment]
So I've been trying to break my uboot patch all day, and it seems to be handling anything I throw at it just fine. @samueldr any idea on how/whether to get this upstream?

**Comment by samueldr** [comment]
The good contributors at U-Boot were, in my experience a year or two back, really receptive at good questions about contribution on their IRC channel. I expect it would still be the same.

Maybe drop in asking for thoughts on the implementation.

As for contributing to U-Boot, it's a linux-like contribution workflow. Though luckily enough there's just one mailing list to consider, not a fractal of them. The change could be sent outright as the initial revision of the change, I've seen worse as initial revisions of changes. Another option is to `[RFC]` the patch, which is not an RFC like in Rust or NixOS, I believe RFCs there stands strictly for "request for comment".

If it was for myself, I would ask on IRC for comments, then send the initial revision without an RFC. It's self-contained to a specific family of devices.

I'll try the change. At a glance, wouldn't this run on "non 4 family" Raspberry Pi devices too? Though with a shallow reading it looks fine, and more importantly, well documented. (Though I wouldn't name drop a distribution in the comment, and rather refer to the [Generic Distro Configuration Concept](https://source.denx.de/u-boot/u-boot/-/blob/master/doc/README.distro).)

**Comment by K900** [comment]
All right, I'll have ask around IRC tomorrow. Also, yes, the comment is mostly for preservation, and the patch does currently apply to all boards (though it seems harmless on older revisions). This is all very much a proof of concept and I will definitely need a real adult to look over this before merging it anywhere.

**Comment by K900** [comment]
Also, the reason I called out NixOS by name there is that most other distros just drop their own DTBs into the firmware partition and let the bootloader handle it. NixOS is an edge case here because it actually has a separate DTB per boot option.

**Comment by samueldr** [comment]
When *actually* following the generic distro configuration concept, other distros *too* will have their FDT replaced by a dtb file from the distro. I'm pretty sure Fedora would have the same issue with their UEFI+U-Boot scheme.

> other distros just drop their own DTBs into the firmware partition

I wouldn't compare non-U-Boot raspberry pi setups, since yes, that would be about right. But U-Boot contributors wouldn't really care about that setup :).

> I'll have to ask around IRC tomorrow

You can ping me if you want to (samueldr on their IRC channel). If you need to chat I'm on the NixOS matrix spaces, [#nixos-on-arm:nixos.org](https://matrix.to/#/#nixos-on-arm:nixos.org) is a good place for that.

**Comment by samueldr** [comment]
Oh, and:

> and the patch does currently apply to all boards (though it seems harmless on older revisions)

The main thing I would be wary about is the fact it adds *bogus warnings* for those non-4 boards. Which will invariably end up being interpreted as worrysome warnings by end-users :).

**Comment by K900** [comment]
I'm expecting the final patch to have some sort of a version check. Possibly even checking the SoC revision and only applying the fixup on C0? Either way that's probably something we should discuss with the uboot maintainers.

**Comment by K900** [comment]
So I started digging through the u-boot mailing list archives and turns out there's a month old patch that does the same thing that I missed somehow on the first go. Here's the patch: https://patchwork.ozlabs.org/project/uboot/patch/20210822143656.289891-1-sjoerd@collabora.com/

I'll update the prebuilt u-boot on my host with it in a few minutes.

**Comment by K900** [comment]
Updated, also pushed a PR adding the existing patch to nixpkgs (see above).
